cmd/cmd.go

```diff
@@ -46,37 +47,119 @@ const newCore = "/snap/core/current"
+// yet because of missing validation or other issues.
+func distroSupportsReExec() bool {
 	if !release.OnClassic {
-		// you're already the real deal, natch
-		return
+		return false
+	}
+	switch release.ReleaseInfo.ID {
+	case "fedora", "centos", "rhel", "opensuse", "suse", "poky":
+		logger.Debugf("re-exec not supported on distro %q yet", release.ReleaseInfo.ID)
+		return false
+	}
+	return true
+}
+
+// coreSupportsReExec returns true if the given core snap should be used as re-exec target.
+//
+// Ensure we do not use older version of snapd, look for info file and ignore
+// version of core that do not yet have it.
+func coreSupportsReExec(corePath string) bool {
+	fullInfo := filepath.Join(corePath, "/usr/lib/snapd/info")
+	if !osutil.FileExists(fullInfo) {
+		return false
+	}
+	content, err := ioutil.ReadFile(fullInfo)
+	if err != nil {
+		logger.Noticef("cannot read snapd info file %q: %s", fullInfo, err)
+		return false
+	}
+	ver := regexp.MustCompile("(?m)^VERSION=(.*)$").FindStringSubmatch(string(content))
+	if len(ver) != 2 {
+		logger.Noticef("cannot find snapd version information in %q", content)
+		return false
+	}
+	// > 0 means our Version is bigger than the version of snapd in core
+	res, err := strutil.VersionCompare(Version, ver[1])
+	if err != nil {
+		logger.Debugf("cannot version compare %q and %q: %s", Version, ver[1], res)
+		return false
+	}
+	if res > 0 {
+		logger.Debugf("core snap (at %q) is older (%q) than distribution package (%q)", corePath, ver[1], Version)
+		return false
 	}
+	return true
+}
 
-	// should we re-exec? no option in the environment means yes
-	if !osutil.GetenvBool(key, true) {
+// InternalToolPath returns the path of the internal snapd tool.
+//
+// The return value is either the path of the tool in the current distribution
+// or in the core snap (or the ubuntu-core snap). This handles spiritual
+// "re-exec" where we run the tool from the core snap if the environment allows
+// us to do so.
+func InternalToolPath(tool string) string {
+	distroTool := filepath.Join(dirs.DistroLibExecDir, tool)
+
+	// If we are asked not to re-execute use distribution packages. This is
+	// "spiritual" re-exec so use the same environment variable to decide.
+	if !osutil.GetenvBool(reExecKey, true) {
 		logger.Debugf("re-exec disabled by user")
-		return
+		return distroTool
 	}
 
-	// can we re-exec? some distributions will need extra work before re-exec really works.
-	switch release.ReleaseInfo.ID {
-	case "fedora", "centos", "rhel", "opensuse", "suse", "poky":
-		logger.Debugf("re-exec not supported on distro %q yet", release.ReleaseInfo.ID)
+	// If the distribution doesn't support re-exec or run-from-core then don't do it.
+	if !distroSupportsReExec() {
+		return distroTool
+	}
+
+	// Is the tool we are after present in the core snap?
+	corePath := newCore
+	coreTool := filepath.Join(newCore, "/usr/lib/snapd", tool)
+	if !osutil.FileExists(coreTool) {
+		corePath = oldCore
```

Comment:
I wanted to keep it as is but I can remove it. If you agree I'll do it in a separate PR (I've started working on tests already)